File: rust/src/parsers/parser.rs
use crate::models::EnvironmentFile;

/// A parser turns the raw text of one supported file format into envdoctor's
/// normalized `EnvironmentFile`. Parsers are independent and format-specific:
/// they know nothing about detectors, the audit, or each other.
///
/// `match` decides whether a file path belongs to this format; `parse` handles
/// the content. A parser must never throw on malformed input — it should
/// produce whatever it can and let the caller report parse errors.
pub trait Parser: Send + Sync {
    fn id(&self) -> &'static str;
    fn match_path(&self, file_path: &camino::Utf8Path) -> bool;
    fn parse(&self, content: &str, file_path: &camino::Utf8Path) -> EnvironmentFile;
}

/// Ordered list of parsers, most specific first.
pub type ParserRegistry = Vec<Box<dyn Parser>>;

/// Match a path against a registry; returns the first parser that claims it.
pub fn parser_for_path<'a>(registry: &'a ParserRegistry, _file_path: &camino::Utf8Path) -> Option<&'a dyn Parser> {
    registry.iter().find(|p| p.match_path(_file_path)).map(|b| b.as_ref())
}

File: dart/lib/src/parsers/parser.dart
import '../models/environment_file.dart';
import 'docker_compose_parser.dart';
import 'env_parser.dart';
import 'github_actions_parser.dart';
import 'k8s_parser.dart';
import 'source_parser.dart';

/// A parser turns the raw text of one supported file format into envdoctor's
/// normalized `EnvironmentFile`. A parser must never throw on malformed input.
abstract class Parser {
  String get id;
  bool matchPath(String filePath);
  EnvironmentFile parse(String content, String filePath);
}

/// Ordered list of parsers, most specific first.
List<Parser> defaultRegistry(List<String> sourceExtensions) => [
      EnvParser(),
      DockerComposeParser(),
      GithubActionsParser(),
      K8sParser(),
      SourceParser(sourceExtensions),
    ];

/// Match a path against a registry; returns the first parser that claims it.
Parser? parserForPath(List<Parser> registry, String filePath) {
  for (final p in registry) {
    if (p.matchPath(filePath)) return p;
  }
  return null;
}
